slog.go

```diff
@@ -0,0 +1,81 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"log/slog"
+)
+
+// SlogImplementor is an interface that a logr.LogSink can implement
+// to support efficient logging through the slog.Logger API.
+type SlogImplementor interface {
```

Comment:
See other comments ("avoid glue code").

Reply:
Can we maybe do this as 2 (or more) PRs?

1) Add trvially obvious bridge support with known gotchas.
2) Make changes to logr to make cotchas less.
3) Add optional interface indicating direct support for slog.

Reply:
We can definitely start with step 1.

Do you want to do that here on some experimental branch or in a separate repo? I'm leaning towards doing it here, in the main package, because I expect that we'll need that eventually. We just shouldn't merge it into "master" yet.

I'm less sure about the order of the next two steps. Solving some of these issues will be hard and take time. The end result will still be less efficient. It would be simpler to do step 3 first and enhance klog and zapr. Then Kubernetes will be ready to support packages using slog as API much sooner.

Reply:
I just saw your comment below about context handling with the key defined in an internal package. That should work, so let me take back the "in the main package": let's start in this repo with "slogr" as package?

Reply:
I'm OK to skip step 2 if we decide step 3 is the better answer.

This branch seems fine, and we can commit things and then revise, as long as we don't tag a release yet.

WRT which package - `logr.ToSlog` and `logr.FromSlog` are pleasant.  I think a `slogr` package is better if the API is really parallel to funcr, zapr, etc - `slogr.New(slog.Handler)`.

Maybe step 1 is to evaluate that difference.

Reply:
Instead of rewriting the entire branch for this PR, I created a new one for step 1: https://github.com/go-logr/logr/pull/205